src/relax/training/utils.cc

```diff
@@ -0,0 +1,167 @@
+    builder_->EmitNormalized(VarBinding(new_var, new_value));
+  }
+
+  // remap orignal dataflow var
+  // TODO(chaofan): a better way to check whether new_ret_var should be dataflow
+  void RemapToDataflow(SeqExpr body) {
+    for (BindingBlock block : body->blocks) {
+      for (Binding binding : block->bindings) {
+        const auto* binding_node = binding.as<VarBindingNode>();
+        if (binding_node && !binding_node->var->IsInstance<DataflowVarNode>()) {
+          Var new_binding_var = DataflowVar(
+              binding_node->var->vid, GetStructInfo(binding_node->var), binding_node->var->span);
+          this->var_remap_[binding_node->var->vid] = new_binding_var;
+        }
+      }
+    }
+  }
+
+  Array<Var> RemapLossParams(const Array<Var>& loss_func_params, Array<Var> new_params) {
+    for (int i = 0; i < static_cast<int>(loss_func_params.size()); ++i) {
+      Var loss_param = loss_func_params[i];
+      if (i < static_cast<int>(orig_rets.size())) {
+        // map return value to loss param
+        if (const auto* var_node = orig_rets[i].as<VarNode>()) {
+          ICHECK(orig_rets[i].as<DataflowVarNode>());
+          orig_rets_var_.push_back(NullOpt);
+          this->var_remap_[loss_param->vid] = GetRef<Var>(var_node);
+        } else {
+          Var new_ret_var =
+              DataflowVar(/*name_hint=*/"ret_" + std::to_string(i), GetStructInfo(orig_rets[i]));
+          orig_rets_var_.push_back(new_ret_var);
+          this->var_remap_[loss_param->vid] = new_ret_var;
+        }
+      } else {
+        // append to the param list
+        Var new_loss_param = Var(loss_param->vid, GetStructInfo(loss_param), loss_param->span);
+        this->var_remap_[loss_param->vid] = new_loss_param;
+        new_params.push_back(new_loss_param);
+      }
+    }
+    return new_params;
+  }
+
+  Array<Expr> orig_rets;
+
+ private:
+  SeqExpr loss_body_;
+  Array<Optional<Var>> orig_rets_var_;
+};
+
+/*!
+ * \brief Local helper to append a specified loss function after the original function.
+ * \param orig_func The function to be appended.
+ * \param loss_func The loss function.
+ * \return The result function after appended.
+ */
+Function AppendLoss(Function orig_func, Function loss_func) {
+  CHECK(orig_func->body->IsInstance<SeqExprNode>())
+      << "the body of the original function is not SeqExpr.";
+  CHECK(loss_func->body->IsInstance<SeqExprNode>())
+      << "the body of the loss function is not SeqExpr.";
+
+  auto param_copied_func = CopyWithNewParams(orig_func);
+  auto seq_expr = Downcast<SeqExpr>(param_copied_func->body);
+
+  AppendLossMutator mutator(Downcast<SeqExpr>(loss_func->body));
+  mutator.RemapToDataflow(seq_expr);
+  // Get the orignal rets. If it is a Tuple, unpack it.
+  if (orig_func->ret_struct_info.as<TupleStructInfoNode>()) {
+    const auto* tuple_node = seq_expr->body.as<TupleNode>();
+    ICHECK(tuple_node != nullptr);
+    for (Expr field : tuple_node->fields) {
```

Comment:
Always prefer const reference if possible
```suggestion
    for (const Expr& field : tuple_node->fields) {
```